### ISSUE TITLE
Custom callback can't write to `logs`

### ISSUE THREAD

**Issue by AshesITR** [rstudio/keras#1230]
Logs is given to `keras::KerasCallback` implementations as an R `list()` instead of a python `dict`.
This means it's not possible to write to `logs` in order to, e.g., log a custom metric in a callback.

Such a feature is used in `callback_reduce_lr_on_plateau()` where the python implementation tracks `lr`:

https://github.com/tensorflow/tensorflow/blob/a4dfb8d1a71385bd6d122e4f27f86dcebb96712d/tensorflow/python/keras/callbacks.py#L2663

I would like to write a custom learning rate callback that has the same feature.

**Comment by t-kalinowski** [comment]
Thank you very much for the pull request! I am hesitant to accept it in it's current state however because it's a potentially breaking change for all the custom callbacks already written that aren't expecting the return object from the R functions to be meaningful. It would be quite a disruptive change.

Also, this style of writing callbacks is about to be deprecated in favor of a more general subclassing usage pattern where you write R6classes that inherit from a python type directly. It's still in the works so not being advertised yet, but the new approach will give you flexibility to extend to all sorts of python objects (e.g., layers, models, constraints, optimizers ...)

I just opened a PR that makes sure you can modify a python dictionary in-place using the new style. (There are some unrelated tests failing that I need to track down before I merge into master) 

With the new approach, you can now do:

```R
CustomMetric <- R6::R6Class(
  "CustomMetric",
  inherit = keras$callbacks$Callback,
  public = list(
    on_epoch_end = function(epoch, logs = NULL) {
      # `epoch` and `logs` will be a python objects here
      # you can use logs$update() or similar
      # if you want to do R operations on `epoch`, call `py_to_r(epoch)`
      logs[["my_epoch"]] <- epoch
      # return value is still ignored
    }
  )
)
CustomMetric <- r_to_py(CustomMetric, convert = FALSE)
cm <- CustomMetric()

model <- ...
model %>% fit(..., callbacks = list(cm))
```

Please try it out and let me know if you run into any issues.

**Comment by t-kalinowski** [comment]
Hi, this is merged in master now. Please let me know if it fixes your issue. Feel free to reopen if you encounter additional issues.

**Comment by AshesITR** [comment]
Thanks, I'll try it out.
